### PR TITLE
rune: support to IAS remote attestation when enclave runtime started and rune attest command.

### PR DESCRIPTION
**Pull request by hustliyilin**
1. Add GetSgxReport PAL API.
2. Support to IAS remote attestation when Enclave Runtime started.
3. Support `rune attest` command.

Signed-off-by: Yilin Li <YiLin.Li@linux.alibaba.com>

### CODE REVIEW

File: rune/libcontainer/configs/enclave.go [hunk 1]
```diff
@@ -7,7 +7,11 @@ const (
...
 
 type Enclave struct {
-	Type string `json:"type"`
-	Path string `json:"path"`
-	Args string `json:"args,omitempty"`
+	Type              string `json:"type"`
+	Path              string `json:"path"`
+	Args              string `json:"args,omitempty"`
+	RaType            string `json:"ra_type,omitempty"`
+	RaSpid            string `json:"ra_spid,omitempty"`
```
Comment (jiazhang0) [inline rune/libcontainer/configs/enclave.go:14]: The below 3 parameters are dedicated for EPID-based RA. So name these variables as RaEpidSpid, RaEpidSubscriptionKey and RaEpidQuoteType.

File: rune/libenclave/configs/config.go [hunk 2]
```diff
@@ -1,7 +1,11 @@
...
 
 type InitEnclaveConfig struct {
-	Type string   `json:"type"`
-	Path string   `json:"path"`
-	Args string   `json:"args"`
+	Type              string `json:"type"`
+	Path              string `json:"path"`
+	Args              string `json:"args"`
+	RaType            string `json:"ra_type"`
+	RaSpid            string `json:"ra_spid"`
```
Comment (jiazhang0) [inline rune/libenclave/configs/config.go:8]: Same way to handle these naming.

File: rune/libcontainer/specconv/spec_linux.go [hunk 3]
```diff
@@ -332,11 +332,35 @@ func createEnclaveConfig(spec *specs.Spec, config *configs.Config) {
...
 	}
 
+	ra_type := filterOut(env, "ENCLAVE_RA_TYPE")
+	if ra_type == "" {
+		ra_type = libcontainerUtils.SearchLabels(config.Labels, "ra_type")
+	}
+
+	ra_spid := filterOut(env, "ENCLAVE_RA_EPID_SPID")
+	if ra_spid == "" {
+		ra_spid = libcontainerUtils.SearchLabels(config.Labels, "ra_spid")
```
Comment (jiazhang0) [inline rune/libcontainer/specconv/spec_linux.go:342]: The annotations are also needed to be renamed as ra_epid_spid.

File: rune/libenclave/intelsgx/aesmd.go [hunk 4]
```diff
@@ -328,7 +328,7 @@ func GetQuote(report []byte, spid string, linkable bool) ([]byte, error) {
 	}
 
 	quote := resp.GetQuote.GetQuote()
-	if len(quote) < QuoteLength || len(quote) != SgxMaxQuoteLength {
+	if len(quote) < QuoteLength || len(quote) > SgxMaxQuoteLength {
```
Comment (jiazhang0) [inline rune/libenclave/intelsgx/aesmd.go:331]: Why changing to `>`? Even the requested bufSize is longer than the actual size, the return value is still equal to the requested one.

File: rune/libenclave/intelsgx/aesmd.go [hunk 5]
```diff
@@ -355,5 +355,6 @@ func GetQuote(report []byte, spid string, linkable bool) ([]byte, error) {
 	logrus.Debugf("  Signature Length:                     %d\n",
 		q.SigLen)
 
-	return resp.GetQuote.GetQuote(), nil
+	validQuote := quote[0 : q.SigLen+QuoteLength]
```
Comment (jiazhang0) [inline rune/libenclave/intelsgx/aesmd.go:358]: This fix should be made in a separate commit.

File: rune/libenclave/intelsgx/aesmd.go (hunk 4 above)
Comment (hustliyilin) [reply rune/libenclave/intelsgx/aesmd.go:331]: > Why changing to `>`? Even the requested bufSize is longer than the actual size, the return value is still equal to the requested one.

Because the bufSize in the `AesmServiceRequest_GetQuote` request is equal to `SgxMaxQuoteLength`, so the return length of Quote can't be greater than  `SgxMaxQuoteLength`.

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 6]
```diff
@@ -1,12 +1,20 @@
...
+	"github.com/opencontainers/runc/libenclave/attestation"
+	_ "github.com/opencontainers/runc/libenclave/attestation/sgx/ias"
+	"github.com/opencontainers/runc/libenclave/intelsgx"
+	"log"
 	"os"
 )
 
 const (
-	palApiVersion = 2
+	palApiVersion = 3
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:17]: Don't set this. Currently pal api v3 is still in draft not released.

File: rune/libenclave/runelet.go [hunk 7]
```diff
@@ -68,8 +68,16 @@ func StartInitialization(cmd []string, cfg *RuneletConfig) (exitCode int32, err
 		}
 
 		// Launch a remote attestation to the enclave runtime.
-		if err = rt.LaunchAttestation(); err != nil {
-			return 1, err
+		product := "development"
+		quote_type := "SGX_UNLINKABLE_SIGNATURE"
+		if config.RaEpidQuoteType == "linkable" {
```
Comment (jiazhang0) [inline rune/libenclave/runelet.go:73]: Please change config.RaEpidQuoteType to integer and use QuoteSignatureTypeUnlinkable or QuoteSignatureTypeLinkable defined in intelsgx package.

File: rune/libenclave/runelet.go [hunk 8]
```diff
@@ -68,8 +68,16 @@ func StartInitialization(cmd []string, cfg *RuneletConfig) (exitCode int32, err
...
+			quote_type = "SGX_LINKABLE_SIGNATURE"
```
Comment (jiazhang0) [inline rune/libenclave/runelet.go:74]: So this variable is not necessary.

File: rune/libenclave/runelet.go [hunk 9]
```diff
@@ -68,8 +68,16 @@ func StartInitialization(cmd []string, cfg *RuneletConfig) (exitCode int32, err
...
+		}
+
+		if config.RaType == "EPID" && config.RaEpidSpid != "" && config.RaEpidSubscriptionKey != "" {
+			if err = rt.LaunchAttestation(config.RaEpidSpid, quote_type, product, config.RaEpidSubscriptionKey); err != nil {
```
Comment (jiazhang0) [inline rune/libenclave/runelet.go:78]: The type of quote_type would be integer not string. The type of variable product should be bool.

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 10]
```diff
@@ -58,3 +62,79 @@ func (pal *enclaveRuntimePal) Destroy() error {
 	api := &enclaveRuntimePalApiV1{}
 	return api.destroy()
 }
+
+func (pal *enclaveRuntimePal) GetSgxReport(targetInfo []byte, data []byte) ([]byte, error) {
+	api := &enclaveRuntimePalApiV3{}
+	return api.GetSgxReport(targetInfo, data)
+}
+
+func parseAttestParameters(Spid string, Product string, Subscription_key string) map[string]string {
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:71]: Please chagnge

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 11]
```diff
@@ -58,3 +62,79 @@ func (pal *enclaveRuntimePal) Destroy() error {
 	api := &enclaveRuntimePalApiV1{}
 	return api.destroy()
 }
+
+func (pal *enclaveRuntimePal) GetSgxReport(targetInfo []byte, data []byte) ([]byte, error) {
+	api := &enclaveRuntimePalApiV3{}
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:67]: Add the following condition: 
```
if pal.version >= 3 {
   balabla...
}
```

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 12]
```diff
@@ -58,3 +62,79 @@ func (pal *enclaveRuntimePal) Destroy() error {
...
+	if pal.GetSgxReport != nil {
+		data := make([]byte, intelsgx.ReportLength)
+
+		targetInfo, err := intelsgx.GetQeTargetInfo()
+		if err != nil || len(targetInfo) != intelsgx.TargetinfoLength {
+			if err == nil {
+				return fmt.Errorf("len(targetInfo) is not %d, but %d", intelsgx.TargetinfoLength, len(targetInfo))
+			}
+			return err
+		}
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:91]: Change to:

```
if err != nil {
  return err
}

if len(targetInfo) != intelsgx.TargetinfoLength {
  return fmt.Errorf("len(targetInfo) is not %d, but %d", intelsgx.TargetinfoLength, len(targetInfo))
}
```

File: rune/libenclave/internal/runtime/pal/api_linux_v3.go [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+		report := make([]byte, reportBufSize)
+		var pTargetInfo unsafe.Pointer = nil
+		var pData unsafe.Pointer = nil
+		if len(targetInfo) > 0 {
+			pTargetInfo = unsafe.Pointer(&targetInfo[0])
+		}
+		if len(data) > 0 {
+			pData = unsafe.Pointer(&data[0])
+		}
+		ret = C.pal_get_reportV3(sym, pTargetInfo,
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/api_linux_v3.go:41]: Add an empty line here.

File: rune/libenclave/internal/runtime/pal/api_linux_v3.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+			C.int(len(targetInfo)),
+			pData,
+			C.int(len(data)),
+			unsafe.Pointer(&report[0]),
+			(*C.int)(unsafe.Pointer(&reportBufSize)))
+
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/api_linux_v3.go:47]: Skip this empty line.

File: rune/libenclave/internal/runtime/pal/api_linux_v3.go [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+		if ret == 0 {
+			return report, nil
+		}
+
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/api_linux_v3.go:51]: ditto

File: rune/libenclave/internal/runtime/pal/api_linux_v3.go (hunk 13 above)
Comment (jiazhang0) [reply rune/libenclave/internal/runtime/pal/api_linux_v3.go:41]: And the suffix V3 is not necessary.

File: rune/libenclave/internal/runtime/pal/api_linux_v3.go [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+	"fmt"
+	"github.com/opencontainers/runc/libcontainer/nsenter"
+	"github.com/opencontainers/runc/libenclave/intelsgx"
+	"unsafe"
+)
+
+type enclaveRuntimePalApiV3 struct {
+}
+
+func (pal *enclaveRuntimePalApiV3) GetSgxReport(targetInfo []byte, data []byte) ([]byte, error) {
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/api_linux_v3.go:26]: Rename the function name to GetLocalReport

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 17]
```diff
@@ -58,3 +62,79 @@ func (pal *enclaveRuntimePal) Destroy() error {
...
+	p["service-class"] = Product
+	p["spid"] = Spid
+	p["subscription-key"] = Subscription_key
+	return p
+}
+
+func (pal *enclaveRuntimePal) Attest(Spid string, Quote_type string, Product string, Subscription_key string) (err error) {
+	api := &enclaveRuntimePal{}
+	if pal.GetSgxReport != nil {
+		data := make([]byte, intelsgx.ReportLength)
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:83]: data is enclave-specific private data so you don't need to set it from pal api.

File: rune/libenclave/internal/runtime/pal/api_linux_v3.go [hunk 18]
```diff
@@ -0,0 +1,58 @@
+package enclave_runtime_pal // import "github.com/opencontainers/runc/libenclave/internal/runtime/pal"
+
+/*
+#include <stdlib.h>
+#include <errno.h>
+
+static int pal_get_reportV3(void *sym, void *target_info, int target_info_len,
+			void *data, int data_len, void *report, int* report_len)
+{
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/api_linux_v3.go:9]: You are assuming sgx sdk get report api as the direct underlaying function but it is not true. Data is enclave-specific data structure and it should be filled inside enclave. So we don't need to pass data parameter. Again, you cannot assume get report api from sgx sdk is the only underlaying function.

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 19]
```diff
@@ -58,3 +62,79 @@ func (pal *enclaveRuntimePal) Destroy() error {
...
+		if err = svc.Check(quote); err != nil {
+			log.Fatal(err)
+			return err
+		}
+
+		status := svc.Verify(quote)
+		if status.ErrorMessage != "" {
+			return fmt.Errorf("%s", status.ErrorMessage)
+		}
+		svc.ShowStatus(status)
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:135]: Add en empty line here.

File: rune/libenclave/internal/runtime/pal/pal_linux.go (hunk 10 above)
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:71]: Rename the variable Subscription_key to subescriptionKey. Also, other parameters should not be capital letter.

File: rune/libcontainer/nsenter/enclave.c [hunk 20]
```diff
@@ -27,6 +27,7 @@ void *fptr_pal_init;
 void *fptr_pal_exec;
 void *fptr_pal_kill;
 void *fptr_pal_destroy;
+void *fptr_pal_GetSgxReport;
```
Comment (jiazhang0) [inline rune/libcontainer/nsenter/enclave.c:30]: Rename to fptr_pal_get_local_report

File: rune/libenclave/runelet.go (hunk 7 above)
Comment (hustliyilin) [reply rune/libenclave/runelet.go:73]: > Please change config.RaEpidQuoteType to integer and use QuoteSignatureTypeUnlinkable or QuoteSignatureTypeLinkable defined in intelsgx package.

 The `config.RaEpidQuoteType` located in `Spec.annotations`， the type of `Spec.annotations` is `map[string]string`. If given a integer value to `config.RaEpidQuoteType`, there would be an wrong "cannot unmarshal number into Go struct field Spec.annotations of type string"

Comment (hustliyilin) [reply rune/libenclave/runelet.go:73]: > > Please change config.RaEpidQuoteType to integer and use QuoteSignatureTypeUnlinkable or QuoteSignatureTypeLinkable defined in intelsgx package.
> 
> The `config.RaEpidQuoteType` located in `Spec.annotations`， the type of `Spec.annotations` is `map[string]string`. If given a `integer` value to `config.RaEpidQuoteType`, there would be an wrong "cannot unmarshal number into Go struct field Spec.annotations of type string"

If we use a function like `strconv.ParseUint` to transfer the `config.RaEpidQuoteType` to a `integer`, the value of a field of  `Spec.annotations` which isn't a digital string always returns 0 using a function like `strconv.ParseUint`.

File: rune/libenclave/runelet.go (hunk 9 above)
Comment (hustliyilin) [reply rune/libenclave/runelet.go:78]: > The type of quote_type would be integer not string. The type of variable product should be bool.

Add `product` to an optional field `Spec.annotations` of config.json is OK, but it must be `string` type.

File: rune/libcontainer/specconv/spec_linux.go [hunk 21]
```diff
@@ -348,6 +348,20 @@ func createEnclaveConfig(spec *specs.Spec, config *configs.Config) {
...
+	}
+	var ra_environment bool
+	if raEnvironment == "Development" || raEnvironment == "dev" {
+		ra_environment = intelsgx.Development_Environment
+	} else if raEnvironment == "Production" || raEnvironment == "product" {
+		ra_environment = intelsgx.Production_Environment
+	} else {
+		logrus.Infof("Unsupported Ra Environment Configuration %v!\n", raEnvironment)
+		return
+	}
```
Comment (jiazhang0) [inline rune/libcontainer/specconv/spec_linux.go:363]: Just collect parameters here.

File: rune/libcontainer/configs/enclave.go [hunk 22]
```diff
@@ -11,6 +11,7 @@ type Enclave struct {
 	Path                  string `json:"path"`
 	Args                  string `json:"args,omitempty"`
 	RaType                bool   `json:"ra_type,omitempty"`
+	RaEnvironment         bool   `json:"ra_environment,omitempty`
```
Comment (jiazhang0) [inline rune/libcontainer/configs/enclave.go:14]: Rename as RaProductEnclave.

File: rune/libcontainer/nsenter/loader.go [hunk 23]
```diff
@@ -41,6 +42,10 @@ func SymAddrPalDestroy() unsafe.Pointer {
 	return unsafe.Pointer(C.fptr_pal_destroy)
 }
 
+func SymAddrPalGetSgxReport() unsafe.Pointer {
```
Comment (jiazhang0) [inline rune/libcontainer/nsenter/loader.go:45]: Rename the function name to SymAddrPalGetLocalReport.

File: rune/libenclave/internal/runtime/pal/api_linux_v3.go [hunk 24]
```diff
@@ -0,0 +1,51 @@
+package enclave_runtime_pal // import "github.com/opencontainers/runc/libenclave/internal/runtime/pal"
+
+/*
+#include <stdlib.h>
+#include <errno.h>
+
+static int pal_get_report(void *sym, void *target_info, int target_info_len,
+			                      void *report, int* report_len)
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/api_linux_v3.go:8]: Fix the indent.

File: rune/libenclave/internal/runtime/pal/api_linux_v3.go [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+	var ret C.int
+	reportBufSize := int32(intelsgx.ReportLength)
+	sym := nsenter.SymAddrPalGetSgxReport()
+
+	for {
+		report := make([]byte, reportBufSize)
+		var pTargetInfo unsafe.Pointer = nil
+		if len(targetInfo) > 0 {
+			pTargetInfo = unsafe.Pointer(&targetInfo[0])
+		}
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/api_linux_v3.go:36]: Move these lines out of for loop.

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 26]
```diff
@@ -58,3 +62,83 @@ func (pal *enclaveRuntimePal) Destroy() error {
 	api := &enclaveRuntimePalApiV1{}
 	return api.destroy()
 }
+
+func (pal *enclaveRuntimePal) GetLocalReport(targetInfo []byte) ([]byte, error) {
+	if pal.version >= 3 {
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:67]: I think you are unknown to what pal.version means here. pal.version indicates which pal api version is implemented by an enclave runtime PAL. So do you mean the older enclave runtime PAL implement get local report API?

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 27]
```diff
@@ -58,3 +62,83 @@ func (pal *enclaveRuntimePal) Destroy() error {
...
+	p := make(map[string]string)
+
+	p["spid"] = spid
+	p["subscription-key"] = subscriptionKey
+	return p
+}
+
+func (pal *enclaveRuntimePal) Attest(spid string, subscriptionKey string, product uint32, quote_type uint32) (err error) {
+	api := &enclaveRuntimePal{}
+	if pal.GetLocalReport != nil {
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:85]: So if

File: rune/libenclave/runelet.go [hunk 28]
```diff
@@ -68,8 +69,10 @@ func StartInitialization(cmd []string, cfg *RuneletConfig) (exitCode int32, err
 		}
 
 		// Launch a remote attestation to the enclave runtime.
-		if err = rt.LaunchAttestation(); err != nil {
-			return 1, err
+		if config.RaType == sgx.EPID && config.RaEpidSpid != "" && config.RaEpidSubscriptionKey != "" {
```
Comment (jiazhang0) [inline rune/libenclave/runelet.go:75]: Simply checking RaType == sgx.EPID is enough?

File: rune/libenclave/internal/runtime/enclave_runtime.go [hunk 29]
```diff
@@ -56,10 +56,10 @@ func StartInitialization(config *configs.InitEnclaveConfig, logLevel string) (*E
 	return rt, nil
 }
 
-func (rt *EnclaveRuntimeWrapper) LaunchAttestation() error {
+func (rt *EnclaveRuntimeWrapper) LaunchAttestation(spid string, subscription_key string, product uint32, quote_type uint32) error {
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/enclave_runtime.go:59]: Don't use "_" to name the variables.

**Comment by jiazhang0** [comment]
@hustliyilin Split the first commits into 2 parts, and the first one describes the changes to PAL API v3. The remaining logic are described in the second one.

File: rune/libenclave/internal/runtime/pal/pal_linux.go [hunk 30]
```diff
@@ -58,3 +62,83 @@ func (pal *enclaveRuntimePal) Destroy() error {
...
+
+func parseAttestParameters(spid string, subscriptionKey string) map[string]string {
+	p := make(map[string]string)
+
+	p["spid"] = spid
+	p["subscription-key"] = subscriptionKey
+	return p
+}
+
+func (pal *enclaveRuntimePal) Attest(spid string, subscriptionKey string, product uint32, quote_type uint32) (err error) {
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/pal_linux.go:83]: Change quote_type to quoteType.

Comment (jiazhang0) [reply rune/libenclave/internal/runtime/pal/pal_linux.go:83]: I see server variable names are still using _. Change them together.

File: rune/libenclave/internal/runtime/pal/api_linux_v3.go [hunk 31]
```diff
@@ -0,0 +1,46 @@
...
+	"fmt"
+	"github.com/opencontainers/runc/libcontainer/nsenter"
+	"github.com/opencontainers/runc/libenclave/intelsgx"
+	"unsafe"
+)
+
+type enclaveRuntimePalApiV3 struct {
+}
+
+func (pal *enclaveRuntimePalApiV3) get_local_report(targetInfo []byte) ([]byte, error) {
```
Comment (jiazhang0) [inline rune/libenclave/internal/runtime/pal/api_linux_v3.go:26]: Change to getLocalRepor().

File: rune/libenclave/attestation/registry.go [hunk 32]
```diff
@@ -3,7 +3,7 @@ package attestation // import "github.com/opencontainers/runc/libenclave/attesta
 var registry []Registry
 
 type Registry interface {
-	Create(p map[string]string) (*Service, error)
+	Create(p map[string]string, raEnclaveType uint32) (*Service, error)
```
Comment (jiazhang0) [inline rune/libenclave/attestation/registry.go:6]: Why not moving raEnclaveType into p (configuration)?

Comment (jiazhang0) [reply rune/libenclave/attestation/registry.go:6]: Because raEnclaveType currently still belongs to enclave-specific concept. Hiding it under p (config)  is appropriate.

Comment (hustliyilin) [reply rune/libenclave/attestation/registry.go:6]: > Why not moving raEnclaveType into p (configuration)?

Because the type of 'p' is `map[string]string` while the type of `raEnclaveType` is `uint32`. MoreOver, We transfered `raEnclaveType` from `string` to `uint32` before. So I think that in order to place `raEnclaveType` to 'p' to revert 'string' to 'uint32' is not necessary.

File: rune/libenclave/proto/agent-service.proto [hunk 33]
```diff
@@ -12,8 +12,16 @@ message AgentServiceRequest{
         int32 sig = 1;
     }
 
+    message Attest {
+	string spid       	= 1;
+	string subscription_key = 2;
```
Comment (jiazhang0) [inline rune/libenclave/proto/agent-service.proto:17]: Remove "_".

File: rune/libenclave/proto/agent-service.proto [hunk 34]
```diff
@@ -12,8 +12,16 @@ message AgentServiceRequest{
...
+	uint32 product    	= 3;
+	uint32 quote_type       = 4;
```
Comment (jiazhang0) [inline rune/libenclave/proto/agent-service.proto:19]: ditto
